### PR TITLE
Implements fixed-point conversion functions

### LINKED ISSUE #559: FixedPoint conversion functions
# FixedPoint conversion functions

## Conceptual overview

This adds two functions to convert fixed-point representations to and from `Bool[]`. This extracts the underlying logic for `PrepareFxP` and `MeasureFxP` as functions.

### Current status

This functions are only implicitly available and would require qubit allocation in order to access them.

### New and modified functions, operations, and UDTs

```qsharp
namespace Microsoft.Quantum.Convert
{

/// # Summary
/// Computes fixed-point approximation for a double and returns it as `Bool` array.
///
/// # Input
/// ## integerBits
/// Assumed number of integerBits (including the sign big)
/// ## fractionalBits
/// Assumed number of fractionalBits
/// ## value
/// Value to be approximated
function FixedPointAsBoolArray(integerBits : Int, fractionalBits : Int, value : Double) : Bool[] {}

/// # Summary
/// Returns the double value of a fixed-point approximation from of a `Bool` array.
///
/// # Input
/// ## integerBits
/// Assumed number of integerBits (including the sign big)
/// ## bits
/// Bit-string representation of approximated number
function BoolArrayAsFixedPoint(integerBits : Int, bits : Bool[]) : Double {}

/// # Summary
/// Discretizes a double value as a double and returns its approximated representation as a double.
///
/// # Input
/// ## integerBits
/// Assumed number of integerBits (including the sign big)
/// ## fractionalBits
/// Assumed number of fractionalBits
/// ## value
/// Value to be approximated
function DoubleAsFixedPoint(integerBits : Int, fractionalBits : Int, value : Double) : Double {
    return BoolArrayAsFixedPoint(integerBits, fractionalBits, FixedPointAsBoolArray(integerBits, fractionalBits, value));
}

}
```

### Current status

TODO.

### Using proposed changes

TODO.

## Open design questions and considerations

These functions should be used in the implementation of `PrepareFxP` and `MeasureFxP`.

### PR DESCRIPTION
**Pull request by msoeken**
Fixes #559; also rewrites `PrepareFxP` using new function.

### CODE REVIEW

File: Numerics/src/FixedPoint/Convert.qs [hunk 1]
```diff
@@ -5,6 +5,40 @@ namespace Microsoft.Quantum.Convert {
     open Microsoft.Quantum.Arrays;
     open Microsoft.Quantum.Math;
 
+    /// # Summary
+    /// Computes fixed-point approximation for a double and returns it as `Bool` array.
+    ///
+    /// # Input
+    /// ## integerBits
+    /// Assumed number of integerBits (including the sign big)
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:13]:
```suggestion
    /// Assumed number of integer bits (including the sign bit).
```

File: Numerics/src/FixedPoint/Convert.qs [hunk 2]
```diff
@@ -5,6 +5,40 @@ namespace Microsoft.Quantum.Convert {
...
+    /// ## fractionalBits
+    /// Assumed number of fractionalBits
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:15]:
```suggestion
    /// Assumed number of fractional bits.
```

File: Numerics/src/FixedPoint/Convert.qs [hunk 3]
```diff
@@ -5,6 +5,40 @@ namespace Microsoft.Quantum.Convert {
...
+    /// ## value
+    /// Value to be approximated
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:17]:
```suggestion
    /// Value to be approximated.
```

File: Numerics/src/FixedPoint/Convert.qs [hunk 4]
```diff
@@ -13,10 +47,24 @@ namespace Microsoft.Quantum.Convert {
...
     /// Bit-string representation of approximated number
-    internal function BoolArrayAsFixedPoint(integerBits : Int, bits : Bool[]) : Double {
+    function BoolArrayAsFixedPoint(integerBits : Int, bits : Bool[]) : Double {
         let numBits = Length(bits);
         let intPart = (Tail(bits) ? -(1 <<< (numBits - 1)) | 0) + BoolArrayAsInt(Most(bits));
         return IntAsDouble(intPart) / PowD(2.0, IntAsDouble(numBits - integerBits));
     }
 
+    /// # Summary
+    /// Discretizes a double value as a double and returns its approximated representation as a double.
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:57]: I suspect this should be the following, but I might be wrong:
```suggestion
    /// Discretizes a double value as a fixed-point approximation and returns its value as a double.
```

File: Numerics/src/FixedPoint/Convert.qs [hunk 5]
```diff
@@ -5,6 +5,40 @@ namespace Microsoft.Quantum.Convert {
...
+        for idx in 0..numBits - 1 {
+            let intConstant = Floor(rescaledConstant);
+            set rescaledConstant = rescaledConstant / 2.0;
+            mutable currentBit = (intConstant &&& 1) == (sign ? 0 | 1);
+            if keepAdding {
+                set keepAdding = currentBit;
+                set currentBit = not currentBit;
+            }
+            if currentBit {
+                set result w/= idx <- true;
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:35]: You could start with an empty array and append to it `currentBit` on each iteration - the code will be a bit more readable, and you could use `_` as the loop array.

File: Numerics/src/FixedPoint/Convert.qs [hunk 6]
```diff
@@ -13,10 +47,24 @@ namespace Microsoft.Quantum.Convert {
...
         let intPart = (Tail(bits) ? -(1 <<< (numBits - 1)) | 0) + BoolArrayAsInt(Most(bits));
         return IntAsDouble(intPart) / PowD(2.0, IntAsDouble(numBits - integerBits));
     }
 
+    /// # Summary
+    /// Discretizes a double value as a double and returns its approximated representation as a double.
+    ///
+    /// # Input
+    /// ## integerBits
+    /// Assumed number of integerBits (including the sign big)
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:61]:
```suggestion
    /// Assumed number of integer bits (including the sign bit).
```

File: Numerics/src/FixedPoint/Convert.qs [hunk 7]
```diff
@@ -13,10 +47,24 @@ namespace Microsoft.Quantum.Convert {
...
+    /// ## fractionalBits
+    /// Assumed number of fractionalBits
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:63]:
```suggestion
    /// Assumed number of fractional bits.
```

File: Numerics/src/FixedPoint/Convert.qs [hunk 8]
```diff
@@ -13,10 +47,24 @@ namespace Microsoft.Quantum.Convert {
...
+    /// ## value
+    /// Value to be approximated
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:65]:
```suggestion
    /// Value to be approximated.
```

File: Numerics/src/FixedPoint/Convert.qs [hunk 9]
```diff
@@ -13,10 +47,24 @@ namespace Microsoft.Quantum.Convert {
     /// Assumed number of integerBits (including the sign big)
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:47]:
```suggestion
    /// Assumed number of integer bits (including the sign bit).
```

File: Numerics/src/FixedPoint/Convert.qs [hunk 10]
```diff
@@ -13,10 +47,24 @@ namespace Microsoft.Quantum.Convert {
...
     /// ## bits
     /// Bit-string representation of approximated number
```
Comment (tcNickolas) [inline Numerics/src/FixedPoint/Convert.qs:49]:
```suggestion
    /// Bit-string representation of approximated number.
```

**Review by tcNickolas — APPROVED** [review]
Looks good, I left a couple of comments

File: Numerics/src/FixedPoint/Convert.qs (hunk 5 above)
Comment (msoeken) [reply Numerics/src/FixedPoint/Convert.qs:35]: I am afraid that this is not good for performance.

**Pull request merged**